dspace-api/src/main/java/org/dspace/discovery/DiscoverResultIterator.java

```diff
@@ -0,0 +1,144 @@
+/**
+ * The contents of this file are subject to the license and copyright
+ * detailed in the LICENSE and NOTICE files at the root of the source
+ * tree and available online at
+ *
+ * http://www.dspace.org/license/
+ */
+package org.dspace.discovery;
+
+import java.io.Serializable;
+import java.sql.SQLException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+import org.dspace.core.Context;
+import org.dspace.core.ReloadableEntity;
+
+/**
+ * Implementation of {@link Iterator} to iterate over the discover search result.
+ *
+ * @param <T>  the type of the indexed object
+ * @param <PK> the type of the id of the indexed object
+ * @author Luca Giamminonni (luca.giamminonni at 4science.it)
+ * @author Luca Giamminonni (luca.giamminonni at 4science.it)
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class DiscoverResultIterator<T extends ReloadableEntity, PK extends Serializable> implements Iterator<T> {
+
+    private final SearchService searchService;
+
+    private final Context context;
+
+    private final IndexableObject<?, ?> scopeObject;
+
+    private final DiscoverQuery discoverQuery;
+    private final boolean uncacheEntitites;
+    private final int maxResults;
+    private int iteratorCounter;
+    private DiscoverResult currentDiscoverResult;
+    private Iterator<IndexableObject> currentSlotIterator;
+
+    public DiscoverResultIterator(Context context, DiscoverQuery discoverQuery) {
+        this(context, null, discoverQuery, true, -1);
+    }
+
+    public DiscoverResultIterator(Context context, DiscoverQuery discoverQuery, boolean uncacheEntities) {
+        this(context, null, discoverQuery, uncacheEntities, -1);
+    }
+
+    public DiscoverResultIterator(Context context, IndexableObject<?, ?> scopeObject, DiscoverQuery discoverQuery) {
+        this(context, scopeObject, discoverQuery, true, -1);
+    }
+
+    public DiscoverResultIterator(Context context, IndexableObject<?, ?> scopeObject, DiscoverQuery discoverQuery,
+                                  boolean uncacheEntities, int maxResults) {
+
+        this.context = context;
+        this.scopeObject = scopeObject;
+        this.discoverQuery = discoverQuery;
+        this.iteratorCounter = discoverQuery.getStart();
+        this.searchService = SearchUtils.getSearchService();
+        this.uncacheEntitites = uncacheEntities;
+        this.maxResults = maxResults;
+
+        updateCurrentSlotIterator();
+    }
+
+    @Override
+    public boolean hasNext() {
```

Comment:
We should also add JavaDocs to these override methods to describe the overridden behavior